### PR TITLE
Port pcl_passthrough_filter to ROS2

### LINKED ISSUE #5: Port pcl_passthrough_filter_service.cpp to ROS2
Port: https://github.com/uml-robotics/armada_behaviors/blob/main/armada_flexbe_utilities/src/service/pcl_passthrough_filter_service.cpp

### PR DESCRIPTION
**Pull request by Tags01**
# Changes Made
## passthrough_filter_service:
- Introduced class format, moved main functionality to the constructor, and marked data private.
- Added missing header files.
- Added error checking for all input parameters: all parameters must produce a volumn with a
  positive, non-zero, non-NaN size. Node will shutdown on error. This was moved to its own
  function.
- Convert `fromROSMsg` -> `moveFromROSMsg`.
- Preserve TF frame information in the result.
- replace explicit `std::shared_ptr(new TYPE) ` with definition with `std::make_shared<TYPE>()` format.
- Convert parameter namespace to `std::string_view` to follow google style guidelines.

## simple_test_passthrough_filter:
 - Omit explicit test checks. This script is harder to test accurately since its very sensitive to input data. However, it's easy to verify the code works visually which is included in the script. These tests will be replaced anyways with dedicated unit tests so these can be omitted for now.

## Setup
- Works on `g++ 11.4.0`, `clang++ 19`, and `icpx 2024.2.1`.
- Lints Properly.
- No Valgrind and sanitizer checks yet.

## Misc
- `CMakeLists.txt`: Added notes about why the program must not use `-Ofast` or `-ffast-math`
- `CMakeLists.txt`: Added the new files so they can be built
- Added two new launch files. These are mostly the same as with other commits. The default parameters for the bounding box in passthrough_filter default to a `1x1x1 m` cube centered at `(0, 0, 0)` 

# Evidence of Working Code
The darker background is the entire field of view of the camera, while the bright color segments only contain the filtered points. The default settings are used.
ezgif-7-fc5fa396ea
When the arrow is moved up, it falls within the bound of the bottom of the bounding box so the colors appear brighter on the image.

### CODE REVIEW

File: pcl_utility_msgs/srv/PCLPassthroughFilter.srv
Comment (GregoryLeMasurier) [inline pcl_utility_msgs/srv/PCLPassthroughFilter.srv:1]: This is the same as the voxel grid service message - it helps make it easier to read this way (and it's the same way it was done before) but I wonder if we can reduce duplicated messages some way. Let's talk about this with the group next week

File: pcl_utilities/src/passthrough_filter_service.cpp [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+    auto temp_cloud = std::make_shared<pcl::PointCloud<pcl::PointXYZRGB>>();
+
+    std::string frame_id = req->cloud_in.header.frame_id;
+    moveFromROSMsg(req->cloud_in, *temp_cloud);
+
+    pcl::PassThrough<pcl::PointXYZRGB> pass_x;
+    pass_x.setInputCloud(temp_cloud);
+    pass_x.setFilterFieldName("x");
+    pass_x.setFilterLimits(x_min_, x_max_);
+    // pass_x.setFilterLimitsNegative(false);
```
Comment (GregoryLeMasurier) [inline pcl_utilities/src/passthrough_filter_service.cpp]: Do we need the commented out line?

**Review by GregoryLeMasurier — CHANGES_REQUESTED** [review]
Awesome! One small thing to clean up

**Comment by Tags01** [comment]
# Changes
- Added explicit dependencies for `rcl_interfaces`/`rclcpp`. These dependencies were missed because they are always included transitively. Dependencies in the CMakeLists.txt/package.xml are now alphabetical.
- Added explicit overflow checks when converting from double to float. This also resolves warnings in Clang with `-Wimplicit-float-conversions`
- Added comments to the launch script
- Changed out lvalue parameter to rvalue parameter in the test. Beforehand I used `std::move` on an lvalue parameter.

File: pcl_utilities/src/tests/simple_test_passthrough_filter.cpp [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+          get_logger(),
+          "A camera message could not be retrieved within 1 second.");
+        continue;
+      }
+
+      process_point_cloud(std::move(point_cloud_message));
+    }
+  }
+
+  void process_point_cloud(PointCloud2 && point_cloud)
```
Comment (GregoryLeMasurier) [inline pcl_utilities/src/tests/simple_test_passthrough_filter.cpp:89]: Why did we change to pass in an address of an address? Is this best practice? It seems to add overhead where we have to use std::move before each call, where it was handled for the user before

File: pcl_utilities/src/passthrough_filter_service.cpp [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+
+using pcl_utility_msgs::srv::PCLPassthroughFilter;
+constexpr std::string_view g_PARAM_NAMESPACE = "filters.passthrough.";
+
+class PassthroughFilterService : public rclcpp::Node
+{
+private:
+  rclcpp::Service<PCLPassthroughFilter>::SharedPtr passthrough_filter_service_;
+  rcl_interfaces::msg::ParameterDescriptor constraints_;
+  double x_min_;
```
Comment (GregoryLeMasurier) [inline pcl_utilities/src/passthrough_filter_service.cpp:45]: Why did we switch from float to double and then add in a bunch of checks to make sure the conversion happened correctly? Why can't we just use floats?

File: pcl_utilities/src/tests/simple_test_passthrough_filter.cpp (hunk 2 above)
Comment (GregoryLeMasurier) [reply pcl_utilities/src/tests/simple_test_passthrough_filter.cpp:89]: Addressed in person

**Review by GregoryLeMasurier — APPROVED** [review]

File: pcl_utilities/src/tests/simple_test_passthrough_filter.cpp [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+ * The node will crash if invalid input is passed to these parameters.
+ *
+ * The filtered point cloud is then published to the topic
+ *   `test_passthrough_filter_service/passthrough_filtered`
+ *
+ * Usage:
+ *    `ros2 launch pcl_utilities test_passthrough_filter_service.xml point_cloud_topic:=<POINT_CLOUD_TOPIC>`
+ */
+#include <chrono>   // std::chrono::seconds
+#include <cstddef>  // size_t
```
Comment (GregoryLeMasurier) [inline pcl_utilities/src/tests/simple_test_passthrough_filter.cpp]: Not used in file

File: pcl_utilities/src/tests/simple_test_passthrough_filter.cpp [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+#include <cstdint>  // int64_t
```
Comment (GregoryLeMasurier) [inline pcl_utilities/src/tests/simple_test_passthrough_filter.cpp]: Not used in this file

File: pcl_utilities/src/tests/simple_test_passthrough_filter.cpp [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+#include <functional>  // std::bind, std::placeholders
```
Comment (GregoryLeMasurier) [inline pcl_utilities/src/tests/simple_test_passthrough_filter.cpp]: Not used in this file

File: pcl_utilities/src/tests/simple_test_passthrough_filter.cpp [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+#include <ios>  // std::fixed, std::setprecision
```
Comment (GregoryLeMasurier) [inline pcl_utilities/src/tests/simple_test_passthrough_filter.cpp]: Not used in this file

File: pcl_utilities/src/tests/simple_test_passthrough_filter.cpp [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+#include <memory>  // std::make_shared
+#include <sstream>  // std::stringstream
```
Comment (GregoryLeMasurier) [inline pcl_utilities/src/tests/simple_test_passthrough_filter.cpp]: Not used in this file

**Review by GregoryLeMasurier — CHANGES_REQUESTED** [review]
Forgot to check includes - small cleanup

**Review by GregoryLeMasurier — APPROVED** [review]

**Review by flynnbm — APPROVED** [review]
Modifications to the original and overall format looks great

**Pull request merged**